packages/xo-server-load-balancer/src/plan.js

```diff
@@ -276,6 +278,189 @@ export default class Plan {
     return vmsAverages
   }
 
+  // ===================================================================
+  // VCPU pre-positionning helpers
+  // ===================================================================
+
+  async _processVcpuPrepositionning() {
+    const promises = []
+
+    const allHosts = await this._getHosts()
+    if (allHosts.length <= 1) {
+      return
+    }
+    const idToHost = keyBy(allHosts, 'id')
+    const allVms = filter(this._getAllRunningVms(), vm => vm.$container in idToHost)
+    const hostList = this._getVCPUHosts(allHosts, allVms)
+    const idealVcpuPerCpuRatio =
+      hostList.reduce((sum, host) => sum + host.vcpuCount, 0) / hostList.reduce((sum, host) => sum + host.cpuCount, 0)
+
+    debugVcpuBalancing('Try to apply VCPU prepositionning.')
+    debugVcpuBalancing(`VCPU count per host: ${inspect(hostList, { depth: null })}`)
+    debugVcpuBalancing(`Average vCPU per CPU : ${idealVcpuPerCpuRatio}`)
+
+    const vmsAverages = await this._getVmsAverages(allVms, idToHost)
+    const { averages: hostsAverages } = await this._getHostStatsAverages({ hosts: allHosts })
+
+    // 1. Find source host from which to migrate.
+    const sources = sortBy(
+      filter(hostList, host => (host.vcpuCount - 1) / host.cpuCount >= idealVcpuPerCpuRatio),
+      [
+        host => -host.vcpuCount / host.cpuCount,
+        // Find host with the most memory used
+        host => hostsAverages[host.id].memoryFree,
+      ]
+    )
+    debugVcpuBalancing(`Sources : ${inspect(sources, { depth: null })}`)
+
+    for (const sourceHost of sources) {
+      let deltaSource = sourceHost.vcpuCount - sourceHost.cpuCount * idealVcpuPerCpuRatio
+      if (deltaSource < 1) {
+        continue
+      }
+
+      // 2. Find destination host.
+      const destinations = sortBy(
+        filter(hostList, host => host.id !== sourceHost.id && host.vcpuCount < host.cpuCount * idealVcpuPerCpuRatio),
+        [
+          // trying to avoid migrations between pools
+          host => host.poolId === sourceHost.poolId,
+          host => host.vcpuCount / host.cpuCount,
+          host => -hostsAverages[host.id].memoryFree,
+        ]
```

Comment:
I don't think we can consider both for this sort, because identical `memoryFree` or `cpu` values are really unlikely. I can't really tell which is better to consider, as we check that both of these values would not go above their thresholds after migration.
If you think it it better to consider cpu usage over memory for this sorting, we should probably also change other sorting in this function to also use cpu usage.